File: 2 year 2 semester/spo/football-championship-system/backend/app/schemas/ticket.py
from pydantic import BaseModel, validator
from datetime import datetime

class TicketCalculationRequest(BaseModel):
    home_team_id: int
    away_team_id: int
    stadium_id: int
    category: str
    
    @validator('category')
    def validate_category(cls, v):
        valid_categories = ['VIP', 'Standard', 'Economy']
        if v not in valid_categories:
            raise ValueError(f'Category must be one of {valid_categories}')
        return v

class TicketCalculationResponse(BaseModel):
    category: str
    price: float
    base_price: float
    prestige_coefficient: float
    category_multiplier: float

class TicketBase(BaseModel):
    match_id: int
    category: str
    price: float
    
    @validator('category')
    def validate_category(cls, v):
        valid_categories = ['VIP', 'Standard', 'Economy']
        if v not in valid_categories:
            raise ValueError(f'Category must be one of {valid_categories}')
        return v
    
    @validator('price')
    def validate_price(cls, v):
        if v <= 0:
            raise ValueError('Price must be positive')
        return v

class TicketCreate(TicketBase):
    pass

class TicketInDBBase(TicketBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True

class Ticket(TicketInDBBase):
    pass

class TicketWithMatch(TicketInDBBase):
    match: 'MatchWithDetails'
    
    class Config:
        orm_mode = True

# Avoid circular imports
from app.schemas.match import MatchWithDetails
TicketWithMatch.update_forward_refs()

File: 2_year_2_semester/spo/football-championship-system/frontend/src/types/ticket.ts
// frontend/src/types/ticket.ts
import { MatchWithDetails } from './match';

/**
 * Категории билетов.
 */
export type TicketCategory = 'VIP' | 'Standard' | 'Economy';

/**
 * Интерфейс для запроса на расчет стоимости билета.
 * Соответствует схеме `TicketCalculationRequest` в Pydantic.
 */
export interface TicketCalculationRequest {
  home_team_id: number;
  away_team_id: number;
  stadium_id: number;
  category: TicketCategory;
}

/**
 * Интерфейс для ответа с расчетом стоимости билета.
 * Соответствует схеме `TicketCalculationResponse` в Pydantic.
 */
export interface TicketCalculationResponse {
  category: TicketCategory;
  price: number;
  base_price: number;
  prestige_coefficient: number;
  category_multiplier: number;
}

/**
 * Основной интерфейс для билета.
 * Соответствует схеме `Ticket` в Pydantic.
 */
export interface Ticket {
  id: number;
  match_id: number;
  category: TicketCategory;
  price: number;
  created_at: string;
  updated_at: string;
}

/**
 * Интерфейс для билета с детальной информацией о матче.
 * Соответствует схеме `TicketWithMatch` в Pydantic.
 */
export interface TicketWithMatch extends Ticket {
  match: MatchWithDetails;
}

/**
 * Интерфейс для создания нового билета.
 * Соответствует схеме `TicketCreate` в Pydantic.
 */
export interface TicketCreate {
  match_id: number;
  category: TicketCategory;
  price: number;
}